### PR TITLE
Fix the NHWC->NCHW transformation for dynamic weights

### PR DESCRIPTION
**Pull request by mvafin**
Transformation was wrong when weights and data comes from the same input

Jira: #39670

### CODE REVIEW

File: model-optimizer/mo/graph/perm_inputs.py [hunk 1]
```diff
@@ -170,6 +170,22 @@ def transpose(op_node: Node, port_info: str, input_port: int):
...
 
+def transpose_nchw_to_nhwc(op_node: Node, port_info: str, input_port: int):
+    graph = op_node.graph
+    # permutation_data_node = get_node_with_permutation(op_node, port_info)
+    # assert permutation_data_node.has_and_set('permutation'), \
+    #     'Data node "{}" does not have permutation for node {}, port_info "{}".'.format(
+    #         permutation_data_node.id, op_node.id, port_info)
+    # permutation = permutation_data_node.permutation
+    # if len(permutation.perm) == 0:
+    #     return
```
Comment (jane-intel) [inline model-optimizer/mo/graph/perm_inputs.py:181]:
```suggestion
```

File: model-optimizer/mo/graph/perm_inputs.py [hunk 2]
```diff
@@ -170,6 +170,22 @@ def transpose(op_node: Node, port_info: str, input_port: int):
...
+
+    transpose_name = op_node.soft_get('name', op_node.id) + '/Transpose'
+    from mo.front.tf.graph_utils import create_op_with_const_inputs  # avoiding recursive imports
+    transpose = create_op_with_const_inputs(
+        graph, Transpose, {1: int64_array([0, 2, 3, 1])}, {'name': transpose_name, 'override_output_shape': True})
```
Comment (jane-intel) [inline model-optimizer/mo/graph/perm_inputs.py:186]: prepare it for any rank

File: model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py [hunk 3]
```diff
@@ -138,6 +149,25 @@ def find_and_replace_pattern(self, graph: Graph):
...
+        dynamic_in_nodes = dict()
+        for node in dynamic_subgraphs:
+            node_type = node.soft_get('type')
+            if node_type not in ['Const', 'Parameter', 'ShapeOf']:
+                idx_lst = list()
+                for idx in [idx for idx, port in node.in_ports().items() if
+                            not port.disconnected() and port.get_source().node not in dynamic_subgraphs]:
+                    # PermuteAttrs.set_permutation(node.in_node(idx), node,
+                    #                              PermuteAttrs.Permutation(perm=int64_array([0, 2, 3, 1]),
+                    #                                                       inv=int64_array([0, 3, 1, 2])))
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:163]:
```suggestion
```

**Review by jane-intel — CHANGES_REQUESTED** [review]

**Comment by lazarevevgeny** [comment]
We need to have a benchmark run to validate these changes.

File: model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py [hunk 4]
```diff
@@ -172,6 +202,45 @@ def walk_up_from_in_ports_to_out_ports(in_ports: Set[Port], out_ports: Set[Port]
             visited_nodes.add(source_node)
         return visited_ports, visited_nodes
 
+    @staticmethod
+    def walk_up_from_in_ports_to_out_ports_not_going_into_conv_weights(in_ports: Set[Port], out_ports: Set[Port],
+                                                                       visited_ports: Set[Port] = None,
+                                                                       visited_nodes: Set[Node] = None):
+        get_weights_port_index = lambda node: node.weights_index if node.has_valid('weights_index') else 1
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:209]: Describe the logic of this function in the comments.

File: model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py [hunk 5]
```diff
@@ -19,12 +19,15 @@
...
 from extensions.middle.pass_separator import PostMiddleStart
+from mo.front.common.partial_infer.utils import int64_array
 from mo.graph.graph import Graph, Node
+from mo.graph.perm_inputs import PermuteInputs
 from mo.graph.port import Port
 from mo.middle.replacement import MiddleReplacementPattern
+from mo.ops.op import PermuteAttrs
 
 
 class MarkSubGraphsWithCorrectLayout(MiddleReplacementPattern):
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:33]: Should we update the description of this transformation?

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py [hunk 6]
```diff
@@ -172,6 +202,45 @@ def walk_up_from_in_ports_to_out_ports(in_ports: Set[Port], out_ports: Set[Port]
             visited_nodes.add(source_node)
         return visited_ports, visited_nodes
 
+    @staticmethod
+    def walk_up_from_in_ports_to_out_ports_not_going_into_conv_weights(in_ports: Set[Port], out_ports: Set[Port],
+                                                                       visited_ports: Set[Port] = None,
+                                                                       visited_nodes: Set[Node] = None):
+        get_weights_port_index = lambda node: node.weights_index if node.has_valid('weights_index') else 1
+        weighted_layer_type_to_in_weights_port = {
+            'Convolution': get_weights_port_index,
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:211]: What about `ConvolutionBackpropData` and `GroupConvolutionBackpropData`?

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py [hunk 7]
```diff
@@ -186,22 +255,29 @@ def get_ports_and_nodes_on_weights(graph):
 
         # collect all input ports with weights
         weight_ports = set()
+        result_ports = set()
         start_ports = set()
         for node in nodes:
             node_type = node.soft_get('type', 'unknown')
             if node_type not in weighted_types:
-                if node_type in ['Const', 'Parameter', 'ShapeOf']:
+                if node_type in ['Const', 'Parameter', 'ShapeOf', 'ExtractImagePatches']:
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:263]: ?! elaborate more on this change

Comment (mvafin) [reply model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:263]: Yes, so I decided to make ExtractImagePatches as weight path producer. Meaning that if we meat EIP node we make it a start of weight path

Comment (jane-intel) [reply model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:263]: It's not true. Why would we do that?

File: model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py (hunk 6 above)
Comment (jane-intel) [reply model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:211]: They appear later on the back phase

File: model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py (hunk 7 above)
Comment (mvafin) [reply model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:263]: Why not? When we go up from the weight input of Convolution and meet EIP node, in my opinion, it is safe to treat EIP as the start of the weight path, why do you think it is wrong?

**Comment by lazarevevgeny** [comment]
@mvafin , any progress on that?

File: model-optimizer/mo/graph/perm_inputs.py [hunk 8]
```diff
@@ -171,6 +171,24 @@ def transpose(op_node: Node, port_info: str, input_port: int):
     op_node.in_port(input_port).get_connection().insert_node(transpose)
 
 
+def transpose_nchw_to_nhwc(op_node: Node, port_info: str, input_port: int):
+    graph = op_node.graph
+    permutation_data_node = get_node_with_permutation(op_node, port_info)
+    rank = len(permutation_data_node.shape)
```
Comment (rkazants) [inline model-optimizer/mo/graph/perm_inputs.py:177]: please use new Graph API

File: model-optimizer/mo/graph/perm_inputs.py [hunk 9]
```diff
@@ -171,6 +171,24 @@ def transpose(op_node: Node, port_info: str, input_port: int):
...
+    assert rank in [4, 5], 'Rank is not 4D or 5D for HCHW to HHWC permutation on node {}.'.format(op_node.id)
+
+    if rank == 4:
+        perm = int64_array([0, 2, 3, 1])
+    else:
+        perm = int64_array([0, 2, 3, 4, 1])
```
Comment (rkazants) [inline model-optimizer/mo/graph/perm_inputs.py:183]: please implement for more common case, it should be easy to do.

File: model-optimizer/mo/graph/perm_inputs.py [hunk 10]
```diff
@@ -171,6 +171,24 @@ def transpose(op_node: Node, port_info: str, input_port: int):
     op_node.in_port(input_port).get_connection().insert_node(transpose)
 
 
+def transpose_nchw_to_nhwc(op_node: Node, port_info: str, input_port: int):
+    graph = op_node.graph
+    permutation_data_node = get_node_with_permutation(op_node, port_info)
```
Comment (rkazants) [inline model-optimizer/mo/graph/perm_inputs.py:176]: please use new graph API

File: model-optimizer/mo/graph/perm_inputs.py [hunk 11]
```diff
@@ -171,6 +171,24 @@ def transpose(op_node: Node, port_info: str, input_port: int):
     op_node.in_port(input_port).get_connection().insert_node(transpose)
 
 
+def transpose_nchw_to_nhwc(op_node: Node, port_info: str, input_port: int):
```
Comment (rkazants) [inline model-optimizer/mo/graph/perm_inputs.py:174]: It does not work properly for my operation `GatherND`, namely, shape prior Transpose is incorrectly computed, if I insert this rule in `shape_infer` function.
May `shape_inference=True` is missed for it?

**Review by rkazants — CHANGES_REQUESTED** [review]
see comments

File: model-optimizer/mo/graph/perm_inputs.py (hunk 11 above)
Comment (rkazants) [inline model-optimizer/mo/graph/perm_inputs.py:174]: Also, as I understant the function is not adjusted to be called using new Graph API, namely, I can access data node but dataaccessor.

File: model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py (hunk 7 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/MarkSubgraphsWithCorrectLayout.py:237]: I agree that the cannot blindly consider "EIP" as a weight path producer. @mvafin , have you analyzed other models with the EIP operation? For example, Yolo models. I want to understand what will happen if the EIP is not on a weight path? Will MO incorrectly convert the model?

Also I see that the MO tests fails. It should also be investigated.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

**Comment by mvafin** [comment]
@lazarevevgeny I had a bug in my implementation which was added after I refactored the code. After fixing the bug model is correct (Multiply shapes are correct) and MO IR Reader check passes.

**Review by lazarevevgeny — APPROVED** [review]

**Pull request merged**